storage/src/mmr/mutable.rs

```diff
@@ -0,0 +1,385 @@
+pub enum Type<V: Array> {
+    /// Indicates the key no longer has a value.
+    Deleted,
+
+    /// Indicates the key now has the wrapped value.
+    Update(V),
+}
+
+/// An operation applied to the store.
+#[derive(Clone)]
+pub struct Operation<K: Array, V: Array> {
+    pub key: K,
+    pub op_type: Type<V>,
+}
+
+/// A mutable key-value store based on an MMR over its log of operations.
+pub struct MutableMmr<K: Array, V: Array, H: CHasher> {
+    /// An MMR over digests of the operations applied to the store. The number of leaves in this MMR
+    /// always equals the number of operations in the `log`.
+    ops: Mmr<H>,
+
+    /// A log of all operations applied to the store in order of occurrence. The index of each
+    /// operation in this vector is called its _location_.
+    ///
+    /// Invariant: An operation's location is always equal to the number of the MMR leaf storing the
+    /// digest of the operation.
+    log: Vec<Operation<K, V>>,
+
+    /// A map from each key to the location in the log containing its most recent update. Only
+    /// contains the keys that currently have a value (that is, deleted keys are not in the map).
+    snapshot: HashMap<K, usize>,
+}
+
+impl<K: Array, V: Array, H: CHasher> Default for MutableMmr<K, V, H> {
+    /// Return a new, empty store.
+    fn default() -> Self {
+        MutableMmr::new()
+    }
+}
+
+impl<K: Array, V: Array, H: CHasher> MutableMmr<K, V, H> {
+    /// Return a new, empty store.
+    pub fn new() -> Self {
+        MutableMmr {
+            ops: Mmr::new(),
+            log: Vec::<Operation<K, V>>::new(),
+            snapshot: HashMap::new(),
+        }
+    }
+
+    /// Return a store initialized to the state corresponding to the operation sequence given by
+    /// `log`.
+    ///
+    /// Performs some checks that the log constitutes a valid sequence of operations, and returns
+    /// [Error::InvalidUpdate] if an invalid sequence is encountered.
+    pub fn init_from_log(hasher: &mut H, log: Vec<Operation<K, V>>) -> Result<Self, Error> {
+        let mut store = MutableMmr {
+            ops: Mmr::new(),
+            log,
+            snapshot: HashMap::new(),
+        };
+
+        for (i, op) in store.log.iter().enumerate() {
+            let digest = Self::op_digest(hasher, op);
+            store.ops.add(hasher, &digest);
+
+            match op.op_type {
+                Type::Deleted => {
+                    let location = store.snapshot.remove(&op.key);
+                    if location.is_none() {
+                        // Shouldn't be allowed to delete a key that already has no value.
+                        error!("deleted key {} not found in snapshot", op.key);
+                        return Err(Error::InvalidUpdate);
+                    }
+                    location
+                }
+                Type::Update(_) => store.snapshot.insert(op.key.clone(), i),
+            };
+        }
+
+        Ok(store)
+    }
+
+    const DELETE_CONTEXT: u8 = 0;
```

Comment:
I would normally suggest we infer from the enum but think it is better to be explicit (to avoid subtle refactor bugs).

Reply:
Alternatively, we could infer from the enum but just hardcode a bunch of roots in the tests.

Reply:
I think keeping it explicit like this may be best since it's part of the specification of the digest.